src/csharp/Grpc.IntegrationTesting/XdsInteropClient.cs

```diff
@@ -0,0 +1,287 @@
+                Console.WriteLine($"RPC {rpcId} succeeded ");
+            }
+            catch (RpcException ex)
+            {
+                statsWatcher.OnRpcComplete(rpcId, null);
+                Console.WriteLine($"RPC {rpcId} failed: {ex}");
+            }
+        }
+
+        private async Task CleanupCompletedTasksAsync(List<Task> tasks)
+        {
+            var toRemove = new List<Task>();
+            foreach (var task in tasks)
+            {
+                if (task.IsCompleted)
+                {
+                    // awaiting tasks that have already completed should be instantaneous
+                    await task;
+                }
+                toRemove.Add(task);
+            }
+            foreach (var task in toRemove)
+            {
+                tasks.Remove(task);
+            }
+        }
+    }
+
+    internal class StatsWatcher
+    {
+        private readonly object myLock = new object();
+        private readonly AtomicCounter rpcIdGenerator = new AtomicCounter(0);
+
+        private long? firstAcceptedRpcId;
+        private int numRpcsWanted;
+        private int rpcsCompleted;
+        private int rpcsNoHostname;
+        private Dictionary<string, int> rpcsByHostname;
+
+        public AtomicCounter RpcIdGenerator => rpcIdGenerator;
+
+        public StatsWatcher()
+        {
+            Reset();
+        }
+
+        public void OnRpcComplete(long rpcId, string responseHostname)
+        {
+            lock (myLock)
+            {
+                if (!firstAcceptedRpcId.HasValue || rpcId < firstAcceptedRpcId || rpcId >= firstAcceptedRpcId + numRpcsWanted)
+                {
+                    return;
+                }
+
+                if (string.IsNullOrEmpty(responseHostname))
+                {
+                    rpcsNoHostname ++;
+                }
+                else 
+                {
+                    if (!rpcsByHostname.ContainsKey(responseHostname))
+                    {
+                        rpcsByHostname[responseHostname] = 0;
+                    }
+                    rpcsByHostname[responseHostname] += 1;
+                }
+                rpcsCompleted += 1;
+
+                if (rpcsCompleted >= numRpcsWanted)
+                {
+                    Monitor.Pulse(myLock);
+                }
+            }
+        }
+
+        public void Reset()
+        {
+            lock (myLock)
+            {
+                firstAcceptedRpcId = null;
+                numRpcsWanted = 0;
+                rpcsCompleted = 0;
+                rpcsNoHostname = 0;
+                rpcsByHostname = new Dictionary<string, int>();
+            }
+        }
+
+        public LoadBalancerStatsResponse WaitForRpcStatsResponse(int rpcsWanted, int timeoutSec)
+        {
+            lock (myLock)
+            {
+                if (firstAcceptedRpcId.HasValue)
+                {
+                    throw new InvalidOperationException("StateWatcher is already collecting stats.");
```

Comment:
It looks like this is set up to only accept one GetClientStats RPC call at a time (not at all familiar with C# so I might be misunderstanding). It seems like it would be simpler and more flexible in the future to instantiate a StatsWatcher per incoming request, rather than having a single StatsWatcher for the entire service that must be reset between GetClientStats RPCs.

Reply:
You're right, but I'm not sure if the approach you're proposing is any simpler:
- you'd still need to lock each StatsWatcher if you wanted to update the stats.
- you'd have to have an additional collection protected by lock ("currently active watchers").
I think one lock is better than two so I'll probably leave as it.

Reply:
Failing to accept two simultaneous RPCs is a little more than just simpler, as it's functionally limited in a way is different from the other languages. It's true that the current tests do not need to send multiple`GetClientStats` RPCs simultaneously, so the test can still pass, but new tests are being added and this is a non-intuitive limitation. The potential downside of this in the future outweighs any drawback of adding one more lock, IMO.

Reply:
Fair enough, I can refactor once the feature is needed, for now it seems that everything passing and it's better to have C# XDS tests running sooner than later.
So for now i think we can leave as is and followup later if needed.